README.rst
repoze.errorlog
===============

.. image:: https://travis-ci.org/repoze/repoze.errorlog.png?branch=master
        :target: https://travis-ci.org/repoze/repoze.errorlog

.. image:: https://readthedocs.org/projects/repozeerrorlog/badge/?version=latest
        :target: http://repozeerrorlog.readthedocs.org/en/latest/
        :alt: Documentation Status

.. image:: https://img.shields.io/pypi/v/repoze.errorlog.svg
        :target: https://pypi.python.org/pypi/repoze.errorlog

.. image:: https://img.shields.io/pypi/pyversions/repoze.errorlog.svg
        :target: https://pypi.python.org/pypi/repoze.errorlog

This package implements a `WSGI <https://www.python.org/dev/peps/pep-0333/>`_
middleware filter which intercepts exceptions and writes them to a Python
logging module channel (or the ``wsgi.errors`` filehandle, if no channel is
configured).  It also allows the browsing of limited exception history via
a browser UI.

Installation
------------

Install using setuptools, e.g. (within a virtualenv)::

 $ easy_install repoze.errorlog

or using pip::

 $ pip install repoze.errorlog


Usage
-----

For details on using the various components, please see the
documentation in ``docs/index.rst``.  A rendered version of that documentation
is also available online:

 - http://repozeerrorlog.readthedocs.org/en/latest/


Reporting Bugs 
--------------

Please report bugs in this package to

  https://github.com/repoze/repoze.errorlog/issues


Obtaining Source Code
---------------------

Download development or tagged versions of the software by visiting:

  https://github.com/repoze/repoze.errorlog



CHANGES.rst
Changelog
=========

1.1.1 (unreleased)
------------------

- TBD

1.1 (2016-06-03)
----------------

- Add support for Python 3.5.

- Drop support for Python 2.6 and 3.2.

1.0.0 (2015-02-07)
------------------

- Add support for Python 3.2, 3,3, and 3.4, and PyPy3.

- Drop dependency on ``Paste`` (forking / simplifying
  ``paste.request.parse_querystring`` and ``paste.request.construct_url``).

- Add support for testing on Travis.

- Add Sphinx documentation.

- Drop support for Python 2.4 / 2.5.

0.9.2 (2012-03-29)
------------------

- This release is the last which will maintain support for Python 2.4 /
  Python 2.5.

- Add support for continuous integration using ``tox`` and ``jenkins``.

- Add support for PyPy.

- Add 'setup.py dev' alias (runs ``setup.py develop`` plus installs
  ``nose`` and ``coverage``).

- Moved to github.

0.9.1 (2010-05-23)
------------------

- Make it possible to send exceptions to a logger channel as described
  in the docs; this didn't actually work previously because a) I'm not
  very good at programming and b) the Python logging module is
  terrible (using ``s[-1]`` and not catching an exception when the
  string is empty rather than using ``s.endswith()``, at least under
  Python 2.4).

0.9 (2010-05-23)
----------------

- Bump copyrights.

- Remove dependency on ``ez_setup.py``.

- Docs now show how to use ``ErrorLog`` outside PasteDeploy.

- Avoid a dependency on `elementtree` when used with Python 2.5 and later.
  In those Python versions we used the built-in ``xml.etree`` support.

- 100% test coverage.

0.8 (2008-06-25)
----------------

- Remove post-mortem debug middleware (moved to ``repoze.debug``).

- Initial PyPI release.

0.7 (2008-05-21)
----------------

- Add post-mortem debug middleware (``egg:repoze.errorlog#pdbpm``)

- Remove versions from dependencies.

0.6
---

- Get rid of find-link point to http://dist.repoze.org in ``setup.py``.

- Bump ``ez_setup.py`` version.

0.5
---

- Depend on ``elementree`` 1.2.6 explicitly.

0.4
---

- Add ``ignore`` feature to configuration.  A value consisting of
  space-separated entry point names can be used here, indicating that
  these exception types should not be logged or kept in exception
  history.

- Allow ``__error_log__`` view path to be configured via ``path`` entry in
  Paste configuration.

- Place ``repoze.errorlog.path`` and ``repoze.errorlog.entryid`` in the
  WSGI environment to allow error-catching middleware and apps to
  compose URLs to errors.

0.3
---

- 0.2 didn't work as a filter. :-(

- Don't use ``zope.pagetemplate``, it has too many (potentially
  conflicting) dependencies.  Instead use ``meld`` for template views.

0.2
---

- Provide a TTW view (accessible via ``/__error_log__``) of recent
  tracebacks.

0.1
---

- Initial release.


repoze/errorlog/tests.py
import sys
import unittest
import logging

class TestTopLevelFuncs(unittest.TestCase):
    def test_make_errorlog_no_filename(self):
        from repoze.errorlog import make_errorlog
        me = sys.modules[__name__].__name__
        dummy = '%s:DummyException' % me
        elog = make_errorlog(None, None, channel='foo', keep=10,
                             path='/__error_log__',
                             ignore='KeyError AttributeError %s' % dummy)
        self.assertEqual(elog.channel, 'foo')
        self.assertEqual(elog.keep, 10)
        self.assertEqual(elog.path, '/__error_log__')
        self.assertEqual(elog.counter, 0)
        self.assertEqual(elog.errors, [])
        self.assertEqual(elog.ignored_exceptions, (KeyError, AttributeError,
                                                   DummyException))

class TestErrorLogging(unittest.TestCase):
    def setUp(self):
        from ._compat import NativeStream
        self.errorstream = NativeStream()
        logging.basicConfig(stream = self.errorstream)

    def tearDown(self):
        from logging import root
        root.handlers = []
        # grr
        from logging import _handlerList
        _handlerList[:] = []
        
    def _makeOne(self, *arg, **kw):
        from repoze.errorlog import ErrorLog
        return ErrorLog(*arg, **kw)

    def test_ctor(self):
        elog = self._makeOne(None, channel='foo', keep=10,
                             path='/__error_log__', ignored_exceptions=())
        self.assertEqual(elog.channel, 'foo')

    def test_log_no_exc(self):
        app = DummyApplication()
        elog = self._makeOne(app, channel='foo', keep=10,
                             path='/__error_log__', ignored_exceptions=())
        env = {}
        result = elog(env, None)
        self.assertEqual(app.environ, env)
        self.assertEqual(app.start_response, None)
        self.assertEqual(result, ['hello world'])
        self.assertEqual(env['repoze.errorlog.path'], '/__error_log__')
        self.assertEqual(env['repoze.errorlog.entryid'], '0')
        
    def test_log_exc_no_channel(self):
        from ._compat import NativeStream
        errors = NativeStream()
        app = DummyApplication(KeyError)
        elog = self._makeOne(app, channel=None, keep=10,
                             path='/__error_log__', ignored_exceptions=())
        env = {'wsgi.errors':errors}
        self.assertRaises(KeyError, elog, env, None)
        self.assertTrue('KeyError' in errors.getvalue())
        self.assertEqual(env['repoze.errorlog.path'], '/__error_log__')
        self.assertEqual(env['repoze.errorlog.entryid'], '0')

    def test_log_exc_with_root_channel(self):
        app = DummyApplication(KeyError)
        root = ''
        elog = self._makeOne(app, channel=root, keep=10,
                             path='/__error_log__', ignored_exceptions=())
        env = {}
        self.assertRaises(KeyError, elog, env, None)
        self.assertTrue('KeyError' in self.errorstream.getvalue())
        self.assertEqual(env['repoze.errorlog.path'], '/__error_log__')
        self.assertEqual(env['repoze.errorlog.entryid'], '0')

    def test_log_ignored_builtin_exceptions(self):
        from ._compat import NativeStream
        errors = NativeStream()
        app = DummyApplication(KeyError)
        env = {'wsgi.errors':errors}
        elog = self._makeOne(app, channel=None, keep=10,
                             path='/__error_log__',
                             ignored_exceptions=(KeyError, AttributeError))
        self.assertRaises(KeyError, elog, env, None)
        self.assertEqual(errors.getvalue(), '')
        self.assertEqual(elog.errors, [])

    def test_identifier_counter(self):
        from ._compat import NativeStream
        errors = NativeStream()
        app = DummyApplication(KeyError)
        elog = self._makeOne(app, channel=None, keep=10,
                             path='/__error_log__', ignored_exceptions=())
        env = {'wsgi.errors':errors}
        self.assertRaises(KeyError, elog, env, None)
        self.assertRaises(KeyError, elog, env, None)
        self.assertRaises(KeyError, elog, env, None)
        self.assertEqual(env['repoze.errorlog.path'], '/__error_log__')
        self.assertEqual(env['repoze.errorlog.entryid'], '2')

    def test_show_index_view(self):
        env = {'PATH_INFO':'/__error_log__', 'wsgi.url_scheme':'http',
               'SERVER_NAME':'localhost', 'SERVER_PORT':'8080'}
        elog = self._makeOne(None, channel=None, keep=10,
                             path='/__error_log__', ignored_exceptions=())
        L = []
        def start_response(code, headers):
            L.append((code, headers))
        from repoze.errorlog import Error
        elog.errors = [
            Error('1','description1','rendering1','time1',env,'url1'),
            Error('2','description2','rendering2','time2',env,'url2')
            ]
        bodylist = elog(env, start_response)
        body = bodylist[0]
        self.assertTrue(b'time1' in body)
        self.assertTrue(b'time2' in body)
        self.assertTrue(b'description1' in body)
        self.assertTrue(b'description2' in body)
        self.assertTrue(b'Recent Errors' in body)

    def test_show_index_view_noerrors(self):
        env = {'PATH_INFO':'/__error_log__', 'wsgi.url_scheme':'http',
               'SERVER_NAME':'localhost', 'SERVER_PORT':'8080'}
        elog = self._makeOne(None, channel=None, keep=10,
                             path='/__error_log__', ignored_exceptions=())
        L = []
        def start_response(code, headers):
            L.append((code, headers))
        elog.errors = []
        bodylist = elog(env, start_response)
        body = bodylist[0]
        self.assertTrue(b'No Recent Errors' in body)
        
    def test_show_entry_view_present(self):
        env = {'PATH_INFO':'/__error_log__', 'wsgi.url_scheme':'http',
               'SERVER_NAME':'localhost', 'SERVER_PORT':'8080',
               'QUERY_STRING':'entry=1'}
        elog = self._makeOne(None, channel=None, keep=10,
                             path='/__error_log__', ignored_exceptions=())
        L = []
        def start_response(code, headers):
            L.append((code, headers))
        from repoze.errorlog import Error
        elog.errors = [
            Error('1','description1','rendering1','time1',env,'url1'),
            ]
        bodylist = elog(env, start_response)
        body = bodylist[0]
        self.assertTrue(b'rendering1' in body)
        self.assertTrue(b'time1' in body)
        self.assertTrue(b'Error ' in body)

    def test_show_entry_view_absent(self):
        env = {'PATH_INFO':'/__error_log__', 'wsgi.url_scheme':'http',
               'SERVER_NAME':'localhost', 'SERVER_PORT':'8080',
               'QUERY_STRING':'entry=1'}
        elog = self._makeOne(None, channel=None, keep=10,
                             path='/__error_log__', ignored_exceptions=())
        L = []
        def start_response(code, headers):
            L.append((code, headers))
        elog.errors = []
        bodylist = elog(env, start_response)
        body = bodylist[0]
        self.assertTrue(b'Error Expired' in body)

    def test_insert_error(self):
        env = {'PATH_INFO':'/__error_log__', 'wsgi.url_scheme':'http',
               'SERVER_NAME':'localhost', 'SERVER_PORT':'8080',
               'QUERY_STRING':'entry=1'}
        elog = self._makeOne(None, channel=None, keep=10,
                             path='/__error_log__', ignored_exceptions=())
        import sys
        try:
            raise KeyError
        except:
            exc_info = sys.exc_info()
        elog.insert_error('id', exc_info, env)
        self.assertEqual(len(elog.errors), 1)

        # rollover
        elog.errors = [None] * 10
        elog.insert_error('id2', exc_info, env)
        self.assertEqual(len(elog.errors), 10)
        from repoze.errorlog import Error
        self.assertEqual(elog.errors[0].__class__, Error)
        del exc_info


class Test__parse_querystring(unittest.TestCase):

    def _callFUT(self, environ):
        from repoze.errorlog import _parse_querystring
        return _parse_querystring(environ)

    def test_no_qs(self):
        environ = {}
        found = self._callFUT(environ)
        self.assertEqual(found, [])

    def test_simple(self):
        environ = {'QUERY_STRING': 'a=1&b=2&c=3'}
        found = self._callFUT(environ)
        self.assertEqual(found,
                         [('a', '1'), ('b', '2'), ('c', '3')])

    def test_w_repeats(self):
        environ = {'QUERY_STRING': 'a=1&b=2&c=3&b=4'}
        found = self._callFUT(environ)
        self.assertEqual(found,
                         [('a', '1'), ('b', '2'), ('c', '3'), ('b', '4')])

    def test_w_empty(self):
        environ = {'QUERY_STRING': 'a&b&c=&d=1'}
        found = self._callFUT(environ)
        self.assertEqual(found, [('a', ''), ('b', ''), ('c', ''), ('d', '1')])


class Test__construct_url(unittest.TestCase):

    def _callFUT(self, environ):
        from repoze.errorlog import _construct_url
        return _construct_url(environ)

    def test_w_HTTP_HOST_wo_port(self):
        environ = _makeEnviron({'HTTP_HOST': 'example.com'})
        self.assertEqual(self._callFUT(environ), 'http://example.com')

    def test_w_HTTP_HOST_w_default_http_port(self):
        environ = _makeEnviron({'HTTP_HOST': 'example.com:80'})
        self.assertEqual(self._callFUT(environ), 'http://example.com')

    def test_w_HTTP_HOST_w_non_default_http_port(self):
        environ = _makeEnviron({'HTTP_HOST': 'example.com:8080'})
        self.assertEqual(self._callFUT(environ), 'http://example.com:8080')

    def test_w_HTTP_HOST_w_default_https_port(self):
        environ = _makeEnviron({'HTTP_HOST': 'example.com:443',
                                'wsgi.url_scheme': 'https',
                               })
        self.assertEqual(self._callFUT(environ), 'https://example.com')

    def test_w_HTTP_HOST_w_non_default_https_port(self):
        environ = _makeEnviron({'HTTP_HOST': 'example.com:4443',
                                'wsgi.url_scheme': 'https',
                               })
        self.assertEqual(self._callFUT(environ), 'https://example.com:4443')

    def test_wo_HTTP_HOST_w_default_http_port(self):
        environ = _makeEnviron({'SERVER_NAME': 'example.com',
                                'SERVER_PORT': '80',
                               })
        self.assertEqual(self._callFUT(environ), 'http://example.com')

    def test_wo_HTTP_HOST_w_non_default_http_port(self):
        environ = _makeEnviron({'SERVER_NAME': 'example.com',
                                'SERVER_PORT': '8080',
                               })
        self.assertEqual(self._callFUT(environ), 'http://example.com:8080')

    def test_wo_HTTP_HOST_w_default_https_port(self):
        environ = _makeEnviron({'SERVER_NAME': 'example.com',
                                'SERVER_PORT': '443',
                                'wsgi.url_scheme': 'https',
                               })
        self.assertEqual(self._callFUT(environ), 'https://example.com')

    def test_wo_HTTP_HOST_w_non_default_https_port(self):
        environ = _makeEnviron({'SERVER_NAME': 'example.com',
                                'SERVER_PORT': '4443',
                                'wsgi.url_scheme': 'https',
                               })
        self.assertEqual(self._callFUT(environ), 'https://example.com:4443')

    def test_w_SCRIPT_NAME(self):
        environ = _makeEnviron({'SERVER_NAME': 'example.com',
                                'SERVER_PORT': '80',
                                'SCRIPT_NAME': '/script/name',
                               })
        self.assertEqual(self._callFUT(environ),
                         'http://example.com/script/name')

    def test_w_PATH_INFO(self):
        environ = _makeEnviron({'SERVER_NAME': 'example.com',
                                'SERVER_PORT': '80',
                                'PATH_INFO': '/path/info',
                               })
        self.assertEqual(self._callFUT(environ),
                         'http://example.com/path/info')

    def test_w_SCRIPT_NAME_and_PATH_INFO(self):
        environ = _makeEnviron({'SERVER_NAME': 'example.com',
                                'SERVER_PORT': '80',
                                'SCRIPT_NAME': '/script/name',
                                'PATH_INFO': '/path/info',
                               })
        self.assertEqual(self._callFUT(environ),
                         'http://example.com/script/name/path/info')


class DummyApplication:
    def __init__(self, exc=None):
        self.exc = exc

    def __call__(self, environ, start_response):
        if self.exc:
            raise self.exc
        self.environ = environ
        self.start_response = start_response
        return ['hello world']

class DummyException(Exception):
    pass


def _makeEnviron(override=None):
    from ._compat import NativeStream
    environ = {
        'SERVER_NAME': 'localhost',
        'SERVER_PORT': '80',
        'wsgi.version': (1, 0),
        'wsgi.multiprocess': False,
        'wsgi.multithread': True,
        'wsgi.run_once': False,
        'wsgi.url_scheme': 'http',
        'wsgi.input': NativeStream('hello world'),
        }
    if override is not None:
        environ.update(override)
    return environ


repoze/errorlog/__init__.py
##############################################################################
#
# Copyright (c) 2007 Agendaless Consulting and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the BSD-like license at
# http://www.repoze.org/LICENSE.txt.  A copy of the license should accompany
# this distribution.  THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL
# EXPRESS OR IMPLIED WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND
# FITNESS FOR A PARTICULAR PURPOSE
#
##############################################################################

from logging import getLogger
import os
import pprint
import sys
import traceback
import time

from ._compat import NativeStream
from ._compat import parse_qsl
from ._compat import quote

import meld3

_HERE = os.path.abspath(os.path.dirname(__file__))

class ErrorLog:
    def __init__(self, application, channel, keep, path, ignored_exceptions):
        """ WSGI Middleware which logs errors to a confligurable place
        and exposes a web user interface to display the last N errors.

        o 'application' is the RHS in the WSGI "pipeline".

        o 'channel' is the logging "channel" (logger name) to send error log
          messages to.

        o 'keep' is the number of errors to keep in exception history for
          through the web viewing.

        o path is the path to the error log view (e.g. '/__error_log__').

        o ignored_exceptions is a list of exceptions (Python references) to
          ignore (to refrain from logging or adding to exception history).
        """
        self.application = application
        self.channel = channel
        self.keep = keep
        self.path = path
        self.counter = 0
        self.ignored_exceptions = ignored_exceptions
        self.errors = []

    def new_identifier(self):
        identifier = str(self.counter)
        self.counter += 1
        return identifier
        
    def __call__(self, environ, start_response):
        path_info = environ.get('PATH_INFO')

        if path_info == self.path:
            # we're being asked to render a view
            querydata = dict(_parse_querystring(environ))
            if 'entry' in querydata:
                body = self.entry(querydata['entry'])
            else:
                url = _construct_url(environ)
                body = self.index(url)
            start_response('200 OK', [('content-type', 'text/html'),
                                      ('content-length', str(len(body)))])
            return [body]
        else:
            # we need to try to catch an error
            identifier = self.new_identifier()
            # we place the error log path and identifier in the
            # environment so the application or other middleware can
            # form a URL to the exception
            environ['repoze.errorlog.path'] = self.path
            environ['repoze.errorlog.entryid'] = identifier
            try:
                return self.application(environ, start_response)
            except self.ignored_exceptions:
                # just reraise an ignored exception
                raise
            except:
                self.insert_error(identifier, sys.exc_info(), environ)
                if self.channel is None:
                    errors = environ.get('wsgi.errors')
                    if errors:
                        traceback.print_exc(None, errors)
                else:
                    logger = getLogger(self.channel)
                    logger.exception('\n')
                raise

    def index(self, url):
        template = os.path.join(_HERE, 'templates', 'errors.html')
        root = meld3.parse_xml(template)
        if self.errors:
            iterator = root.findmeld('error_li').repeat(self.errors)
            for li_element, error in iterator:
                t = li_element.findmeld('error_time')
                t.content(error.time)
                url = li_element.findmeld('error_url')
                url.attributes(href=error.url)
                url.content(error.description)
        else:
            content = root.findmeld('content')
            content.content('<h1>No Recent Errors</h1>', structure=True)
        return root.write_xhtmlstring()

    def entry(self, identifier):
        template = os.path.join(_HERE, 'templates', 'entry.html')
        error = self.get_error(identifier)
        root = meld3.parse_xml(template)
        if error:
            header = root.findmeld('header')
            header.content('Error at %s' % error.time)
            text = root.findmeld('text')
            text.content(error.text)
        else:
            header = root.findmeld('header')
            header.content('Error Expired')
            text_area = root.findmeld('text_area')
            text_area.content(
                "<p>The error you're attempting to view has been flushed from "
                "the online error log history.</p>", structure=True)
        return root.write_xhtmlstring()

    def get_error(self, identifier):
        for error in self.errors:
            if error.identifier == identifier:
                return error

    def insert_error(self, identifier, exc_info, environ):
        if len(self.errors) >= self.keep:
            self.errors.pop()
        f = NativeStream()
        # we can't unpack the exception tuple or we'd cause a cycle
        traceback.print_exception(exc_info[0],exc_info[1],exc_info[2],None,f)
        desc = str(exc_info[0])
        tb_rendering = f.getvalue()
        time_str = time.ctime()
        url = self.path +'?entry=%s' % identifier
        error = Error(identifier, desc, tb_rendering, time_str, environ, url)
        self.errors.insert(0, error)
        
class Error:
    """Capture information about a single exception.
    """
    def __init__(self, identifier, desc, tb_rendering, time, environ, url):
        self.identifier = identifier
        self.description = desc
        self.text = tb_rendering + '\n\n' + pprint.pformat(environ)
        self.time = time
        self.url = url
    
def make_errorlog(app, global_conf, **local_conf):
    """Paste filterapp factory.
    """
    channel = local_conf.get('channel', None)
    keep = int(local_conf.get('keep', 20))
    path = local_conf.get('path', '/__error_log__')
    ignore = local_conf.get('ignore', None)
    # e.g. Paste.httpexceptions.HTTPFound,
    # Paste.httpexceptions.HTTPUnauthorized, Paste.httpexceptions.HTTPNotFound
    ignored_exceptions = []
    if ignore:
        from pkg_resources import EntryPoint
        ignore_names = [ name.strip() for name in ignore.split() ]
        bdict = __builtins__
        if not isinstance(bdict, dict): #pragma NO COVER pypy
            bdict = bdict.__dict__
        for name in ignore_names:
            if name in bdict:
                ignored_exc = bdict[name]
            else:
                ep = EntryPoint.parse('x=%s' % name)
                ignored_exc = EntryPoint.parse('x=%s' % name).resolve()

            ignored_exceptions.append(ignored_exc)
    ignored_exceptions = tuple(ignored_exceptions)
    return ErrorLog(app, channel, keep, path, ignored_exceptions)


def _parse_querystring(environ):
    """Parse a query string into a list like ``[(name, value)]``.

    You can pass the result to ``dict()``, but be aware that keys that
    appear multiple times will be lost (only the last value will be
    preserved).

    Forked / simplified from ``paste.request.parse_querystring`` (to allow
    port to Py3k).
    """
    source = environ.get('QUERY_STRING', '')
    if not source:
        return []
    parsed = parse_qsl(source, keep_blank_values=True, strict_parsing=False)
    return parsed


def _construct_url(environ):
    """Reconstruct the URL from the WSGI environment.

    You may override SCRIPT_NAME, PATH_INFO, and QUERYSTRING with
    the keyword arguments.

    Forked / simplified from ``paste.request.construct_url`` (to allow
    port to Py3k).
    """
    url = environ['wsgi.url_scheme'] + '://'

    if environ.get('HTTP_HOST'):
        host = environ['HTTP_HOST']
        port = None
        if ':' in host:
            host, port = host.split(':', 1)
            if environ['wsgi.url_scheme'] == 'https':
                if port == '443':
                    port = None
            elif environ['wsgi.url_scheme'] == 'http':
                if port == '80':
                    port = None
        url += host
        if port:
            url += ':%s' % port
    else:
        url += environ['SERVER_NAME']
        if environ['wsgi.url_scheme'] == 'https':
            if environ['SERVER_PORT'] != '443':
                url += ':' + environ['SERVER_PORT']
        else:
            if environ['SERVER_PORT'] != '80':
                url += ':' + environ['SERVER_PORT']

    url += quote(environ.get('SCRIPT_NAME',''))
    url += quote(environ.get('PATH_INFO',''))
    return url
